things/postgres/groups.go

```diff
@@ -0,0 +1,647 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+	"github.com/mainflux/mainflux/internal/groups"
+	"github.com/mainflux/mainflux/pkg/errors"
+)
+
+const maxLevel = 5
+
+var (
+	errDeleteGroupDB          = errors.New("delete group failed")
+	errSelectDb               = errors.New("select group from db error")
+	errConvertingStringToUUID = errors.New("error converting string")
+)
+
+var _ groups.Repository = (*groupRepository)(nil)
+
+type groupRepository struct {
+	db Database
+}
+
+// NewGroupRepo instantiates a PostgreSQL implementation of group
+// repository.
+func NewGroupRepo(db Database) groups.Repository {
+	return &groupRepository{
+		db: db,
+	}
+}
+
+func (gr groupRepository) Save(ctx context.Context, g groups.Group) (groups.Group, error) {
+	var id string
+	path := g.Name
+	q := `INSERT INTO thing_groups (name, description, id, owner_id, metadata, path) VALUES (:name, :description, :id, :owner_id, :metadata, :path) RETURNING id`
+	if g.ParentID != "" {
+		q = `INSERT INTO thing_groups (name, description, id, owner_id, parent_id, metadata, path) VALUES (:name, :description, :id, :owner_id, :parent_id, :metadata, :path) RETURNING id`
+		parent, err := gr.RetrieveByID(ctx, g.ParentID)
+		if err != nil {
+			return groups.Group{}, errors.Wrap(groups.ErrMissingParent, err)
+		}
+		path = parent.Path + "." + g.Name
+	}
+	g.Path = path
+
+	dbu, err := toDBGroup(g)
+	if err != nil {
+		return groups.Group{}, err
+	}
+
+	row, err := gr.db.NamedQueryContext(ctx, q, dbu)
+	if err != nil {
+		pqErr, ok := err.(*pq.Error)
+		if ok {
+			switch pqErr.Code.Name() {
+			case errInvalid, errTruncation:
+				return groups.Group{}, errors.Wrap(groups.ErrMalformedEntity, err)
+			case errDuplicate:
+				return groups.Group{}, errors.Wrap(groups.ErrGroupConflict, err)
+			}
+		}
+
+		return groups.Group{}, errors.Wrap(groups.ErrCreateGroup, err)
+	}
+
+	defer row.Close()
+	row.Next()
+	if err := row.Scan(&id); err != nil {
+		return groups.Group{}, err
+	}
+	g.ID = id
+	return g, nil
+}
+
+func (gr groupRepository) Update(ctx context.Context, g groups.Group) (groups.Group, error) {
+	q := `UPDATE thing_groups SET name = :name, description = :description, parent_id = :parent_id, metadata = :metadata  WHERE id = :id`
+
+	old, err := gr.RetrieveByID(ctx, g.ID)
+	if err != nil {
+		return groups.Group{}, errors.Wrap(errUpdateDB, err)
+	}
+	if g.Name == "" {
+		g.Name = old.Name
+	}
+	if g.Description == "" {
+		g.Description = old.Description
+	}
```

Comment:
Consider allowing `name` and `description` to be empty. While it's pointless to have the `name` field empty, it will eliminate this entire fetch from the DB, as well as potential transaction problem we have here (fetching and updating are executed in separate transactions) due to the lack of optimistic lock.